File: Dynamic Programming/maximum_product_subarray.py
class Solution:
    """
    求子序列中乘积最大的那个子序列，使用map(i, j)记录 i,j 子序列的乘积，最后返回最大的乘积
    时间复杂度和空间复杂度都是 O(n*n)
    方法是OK的，但数据一多就 Time limit 了
    """
    # def maxProduct(self, nums):
        # length = len(nums)
        # result = []

        # for i in range(length):
        #     last = nums[i]
        #     result.append(last)
        #     for j in range(i+1, length):
        #         last *= nums[j]
        #         result.append(last)


        # return max(result)

    """
    另一种方法是记下当前最大和最小，如果遇到负数的话，大的变小，小的变大，
    之后再与每次遍历的值作比较，记下当前最小和最大，这样可以同时找序列
    乘积最小和最大的
    时间复杂度为 O(n) 空间复杂度为 O(1)
    """
    def maxProduct(self, nums):
        r = nums[0]
        imax = r
        imin = r

        for num in nums[1:]:
            if num < 0:
                imax, imin = imin, imax

            imax = max(num, imax*num)
            imin = min(num, imin*num)

            r = max(r, imax)

        return r


s = Solution()

print(s.maxProduct([-2,6,4,-1,0,4]))

File: Dynamic Programming/MaximumProductSubarray.java
class MaximumProductSubarray {
    /**
     * 求连续数组的最大乘积，使用双循环的方法时间复杂度为 O(n^2)
     */
    public int maxProduct1(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        
        int result = nums[0], len = nums.length;
        
        for (int i = 0; i < len; i++) {
            int product = nums[i];
            result = Math.max(result, nums[i]);
            for (int j = i+1; j < len; j++) {
                product *= nums[j];
                result = Math.max(result, product);
            }
        }
        
        return result;
        
    }


    /**
     * 使用两个数组记录包含 nums[i] 的最大连续子数组积
     * 以及最小连续子数组积
     */
    public int maxProduct2(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        
        int result = nums[0], len = nums.length;
        
        int[] maxs = new int[len];
        int[] mins = new int[len];
        
        maxs[0] = nums[0];
        mins[0] = nums[0];
        
        for (int i = 1; i < len; i++) {
            maxs[i] = Math.max(Math.max(maxs[i-1] * nums[i], mins[i-1] * nums[i]), nums[i]);
            mins[i] = Math.min(Math.min(maxs[i-1] * nums[i], mins[i-1] * nums[i]), nums[i]);
            
            // 因为有 0 的存在，不能用 maxs[len-1] 来表示最终结果
            result = Math.max(result, maxs[i]);
        }
        
        return result;
        
    }


    /**
     * 优化
     */
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) return 0;

        int result = nums[0], imax = nums[0], imin = nums[0];

        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < 0) {
                int temp = imax;
                imax = imin;
                imin = temp;
            }

            imax = Math.max(nums[i], imax * nums[i]);
            imin = Math.min(nums[i], imin * nums[i]);


            result = Math.max(result, imax);
        }

        return result;
    }

    public static void main(String[] args) {
        MaximumProductSubarray solution = new MaximumProductSubarray();

        solution.maxProduct(new int[]{-2,0,-1,3,-1});

    }
}